### ISSUE TITLE
[CSA] False negative: dereference of a globally initialized pointer beyond an array is missed

### ISSUE THREAD

**Issue by rikki322** [llvm/llvm-project#214220]
Hi, I found a false negative in `security.ArrayBound`.

## Program

```c
int values[10];
int *invalid = values + 11;

int read_invalid(void) {
  return *invalid;
}
```

The global pointer is initialized beyond the end of `values`, and its dereference is invalid. CSA emits no `security.ArrayBound` diagnostic.

## CSA reproduction

```sh
clang --analyze \
  -Xclang -analyzer-checker=security.ArrayBound \
  global_one_past_array_pointer.c
```

## Comparison with related programs

A direct out-of-bounds array access is reported:

```c
int values[10];

int read_direct(void) {
  return values[11]; // warning: out of bound access
}
```

The missed program moves the same constant out-of-bounds position into a global pointer initializer, then dereferences that pointer later:

```c
int values[10];
int *invalid = values + 11;

int read_invalid(void) {
  return *invalid; // no warning
}
```

## Runtime confirmation

AddressSanitizer reports a global-buffer-overflow when the program dereferences the pointer.

## Expected result

The analyzer should preserve the array extent through a constant global pointer initializer and diagnose the later dereference.

## Verification

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by steakhal** [comment]
Nothing guarantees that `invalid` is invalid. I can see another TU that assignes a different value to that global that always happens before `read_invalid` is invoked in my program.

It's a fair point that the initializer expression didn't raise an issue though.

**Comment by steakhal** [comment]
Please create another issue for checking the global variable initializer and close this one or repurpose this issue. Either works.

**Comment by NagyDonat** [comment]
> It's a fair point that the initializer expression didn't raise an issue though.

Creating a checker that reports undefined pointer arithmetic (i.e. when the result is not in bounds and not a past-the-end pointer) is on my TODO list, in fact it is one of the main motivations why I'm refactoring `security.ArrayBound` now.

I intend to implement it as a separate checker (not a part of `security.ArrayBound`) because I think it is reasonable to distinguish forming an invalid pointer (which is undefined behavior, but often harmless in practice) and actually accessing an invalid memory area (which is very dangerous).

Note that we already have [`alpha.core.PointerArithm`](https://clang.llvm.org/docs/analyzer/checkers.html#alpha-core-pointerarithm-c) which reports pointer arithmetic on pointers that do not point to an array (but does not report pointer arithmetic that overflows or underflows an array memory area) and [`alpha.security.ReturnPtrRange`](https://clang.llvm.org/docs/analyzer/checkers.html#alpha-security-returnptrrange-c) which reports situations where an invalid pointer is _returned_ from a function.

I don't know whether I should directly improve and stabilize those alpha checkers, or just introduce their common generalization, a checker that reports all undefined pointer arithmetic. (By the way @steakhal what would you prefer?)

**Comment by steakhal** [comment]
> > It's a fair point that the initializer expression didn't raise an issue though.
> 
> Creating a checker that reports undefined pointer arithmetic (i.e. when the result is not in bounds and not a past-the-end pointer) is on my TODO list, in fact it is one of the main motivations why I'm refactoring `security.ArrayBound` now.
> 
> I intend to implement it as a separate checker (not a part of `security.ArrayBound`) because I think it is reasonable to distinguish forming an invalid pointer (which is undefined behavior, but often harmless in practice) and actually accessing an invalid memory area (which is very dangerous).
> 
> Note that we already have [`alpha.core.PointerArithm`](https://clang.llvm.org/docs/analyzer/checkers.html#alpha-core-pointerarithm-c) which reports pointer arithmetic on pointers that do not point to an array (but does not report pointer arithmetic that overflows or underflows an array memory area) and [`alpha.security.ReturnPtrRange`](https://clang.llvm.org/docs/analyzer/checkers.html#alpha-security-returnptrrange-c) which reports situations where an invalid pointer is _returned_ from a function.
> 
> I don't know whether I should directly improve and stabilize those alpha checkers, or just introduce their common generalization, a checker that reports all undefined pointer arithmetic. (By the way [@steakhal](https://github.com/steakhal) what would you prefer?)

As of today, global initializer expression, such as the following code in the global scope was never the concern of the static analyzer.

```c++
int values[10];
int *invalid = values + 11;
```

IMO if the static analyzer wants to path-sensitively check global initializer, then that needs an RFC and a design discussion. My point was that for this simple case, an AST matcher-based solution would also suffice. In this particular case, probably a builtin compiler diagnostic.

**Comment by NagyDonat** [comment]
> As of today, global initializer expression, such as the following code in the global scope was never the concern of the static analyzer.
> 
> int values[10];
> int *invalid = values + 11;
> 
> IMO if the static analyzer wants to path-sensitively check global initializer, then that needs an RFC and a design discussion.

Oh, I passively knew that global initializer expressions are not checked by the analyzer, but this didn't spring to my mind when I read your remark.

I agree that if we want to support checking these with the static analyzer, then we'd need to start with a design discussion.
